docs/contributor/updateability.md

```diff
@@ -0,0 +1,650 @@
+
+In short, it works by running health check scripts
+(placed in dirs: `/etc/greenboot/check/{required.d,wanted.d}/`) and, depending
+on result of *required* scripts, will run either "green" (healthy) or
+"red" (unhealthy) scripts. 
+
+Greenboot heavily integrates with grub. In fact, it is the grub that decrements
+the `boot_counter` and actually rolls back the system (by selecting the other
+boot target). See [Grub](#grub-ostree) section for more information.
+
+MicroShift delivers these scripts in `microshift-greenboot` RPM package.
+
+Health check script for MicroShift is resides in the repository
+as `packaging/greenboot/microshift-running-check.sh`.
+
+Pre-updateability (and continues to do so), `microshift-greenboot` started to
+provide `packaging/greenboot/microshift-pre-rollback.sh` file as a "red" script.
+If `boot_counter` equals `0`, the script will run `microshift-cleanup-data --ovn`.
+
+> Note: It's a red script, so it runs after healthchecks decided that
+> system is unhealthy, and cleanup only happens when `boot_counter`
+> already fell down to `0`.
+
+### Updateability additions to `microshift-greenboot`
+
+Updateability feature added two, new files - one "green" and one "red" script:
+- `packaging/greenboot/microshift_set_healthy.sh` -> `/etc/greenboot/green.d/40_microshift_set_healthy.sh`
+- `packaging/greenboot/microshift_set_unhealthy.sh` -> `/etc/greenboot/red.d/40_microshift_set_unhealthy.sh`
+
+They work only on ostree-based systems (by checking if `/run/ostree-booted` exists).
+Reason for this is two-fold: how greenboot works on non-ostree systems, and
+deployment ID not being available, leaving only boot ID and health, which
+are not as useful without the deployment as backup name.
+Also, since automated backup management is only intended for ostree systems,
+there is no gain in creating `health.json` (also: information about the health
+can be found in the journal with greater context).
+
+Primary responsibility of "set healthy" and "set unhealthy" scripts
+is to create or update `health.json`.
+
+`microshift_set_unhealthy.sh` has an additional logic that, before
+updating `health.json`, checks if backup matching deployment and boot IDs
+from the `health.json` exists, and does not update the file if not.
+This ensures that, when upgrading (from a healthy system) MicroShift fails to
+create a backup, in effect causing the system to be unhealthy, that the new health
+information does not overwrite existing one ("healthy" instructs MicroShift to
+create a backup).
+Without this mechanism in place, healthy data of previous deployment, would not
+get backed up, `health.json` get updated with `unhealthy` causing subsequent
+starts of MicroShift want to restore a backup (which might not exist because it
+wasn't created, or it's outdated because it contains data of
+"second to last boot of previous deployment").
```

Comment:
Why do I need to know this information? Seems like an implementation detail to me.

Reply:
I'm not sure, it is implementation detail, but this is a contributor's doc.
I think it might be useful to at least mention it because I can imagine someone looking up the file and seeing it has `healthy` but the system rebooted and rolled back, but the greenboot didn't finish yet, so the `health.json` has old info. I can even imagine a bug: "health.json contains healthy, even though system was unhealthy"

Reply:
For the sake of getting all of the "how and why" in one place, let's keep the detail here, for now. If we decide expanding the comments in the script is better we can move some content, but for the short term it's more important to have it written down at all than to get it into exactly the right place.